core/node/crypto/config.go

```diff
@@ -3,624 +3,492 @@ package crypto
+}
 
-	if e.Deleted {
-		occ.settings.Remove(ctx, configKey, e.Block)
-	} else {
-		value, err := configKey.decode(e.Value)
-		if err != nil {
-			log.Error("OnChainConfiguration: received config update with invalid value",
-				"tx", event.TxHash, "key", configKey.name, "err", err)
-			return
+func (occ *onChainConfiguration) GetOnBlock(block BlockNumber) *OnChainSettings {
+	settings := *occ.cfg.Load()
+	// Go in reverse order to find the most recent settings
+	for i := len(settings) - 1; i >= 0; i-- {
+		if block >= settings[i].FromBlockNumber {
+			return settings[i]
 		}
-		occ.settings.Set(ctx, configKey, e.Block, value, e.Value)
 	}
+	// First element should always be the default settings with block number 0.
+	panic("never")
 }
 
-func (occ *onChainConfiguration) GetUint64(key ChainKey) (uint64, error) {
-	blockNum := occ.activeBlock.Load()
-	return occ.GetUint64OnBlock(blockNum, key)
+func (occ *onChainConfiguration) All() []*OnChainSettings {
+	return *occ.cfg.Load()
 }
 
-func (occ *onChainConfiguration) GetInt64(key ChainKey) (int64, error) {
-	blockNum := occ.activeBlock.Load()
-	return occ.GetInt64OnBlock(blockNum, key)
+func (occ *onChainConfiguration) LastAppliedEvent() *river.RiverConfigV1ConfigurationChanged {
+	return occ.lastAppliedEvent.Load()
 }
 
-func (occ *onChainConfiguration) GetInt(key ChainKey) (int, error) {
-	blockNum := occ.activeBlock.Load()
-	return occ.GetIntOnBlock(blockNum, key)
+func HashSettingName(name string) common.Hash {
+	return crypto.Keccak256Hash([]byte(strings.ToLower(name)))
 }
 
-func (occ *onChainConfiguration) GetRawValue(key ChainKey) []byte {
-	blockNum := occ.activeBlock.Load()
-	return occ.GetRawValueOnBlock(blockNum, key)
-}
+func (occ *onChainConfiguration) processRawSettings(
+	ctx context.Context,
+) {
+	log := dlog.FromCtx(ctx)
 
-func (occ *onChainConfiguration) GetMinEventsPerSnapshot(streamType byte) (int, error) {
-	if key, ok := streamTypeToMinEventsPerSnapshotKey[streamType]; ok {
-		if val, err := occ.GetInt(key); err == nil {
-			return val, nil
-		}
+	var blockNums []BlockNumber
+	for key := range occ.loadedSettingMap {
+		blockNums = append(blockNums, key)
 	}
-	return occ.GetInt(StreamDefaultMinEventsPerSnapshotConfigKey)
-}
+	slices.Sort(blockNums)
+
+	// First settings are the default settings
+	settings := []*OnChainSettings{DefaultOnChainSettings()}
 
-func (occ *onChainConfiguration) GetStreamMembershipLimit(streamType byte) (int, error) {
-	if key, ok := streamTypeToUserLimitKey[streamType]; ok {
-		if val, err := occ.GetInt(key); err == nil {
-			return val, err
+	decodeHook := abiBytesToTypeDecoder(ctx)
+	input := make(map[string]any)
+	for _, blockNum := range blockNums {
+		for _, v := range occ.loadedSettingMap[blockNum] {
+			name, ok := occ.keyHashToName[v.key]
+			if !ok {
+				log.Warn("Unknown setting key", "key", v.key, "value", v.value, "block", blockNum)
+				continue
+			}
+			if v.deleted {
+				input[name] = occ.defaultsMap[name]
```

Comment:
What happens in the following scenario:

block 0: `stream.media.maxChunkCount` = 50 (default)
block 5: `stream.media.maxChunkCount` = 100
block 10: `stream.media.maxChunkCount` = 200

200 gives problems and the config change on block 10 is reverted/deleted. It should fall back to 100 but this falls back to the default value of 50?

Reply:
great point!
I looked at the contract code, I need to change processing to match expectations there.

Reply:
now code matches proposed contract changes.